python3Packages.gradio: init at 3.20.1

pbsds opened the pull request:
###### Description of changes

This PR add gradio by @gradio-app, a web server framework for making AI (anything really) demos.

I made this PR some time ago, made upstream fix some issues, then rebased it on master. The shap and markdown-it-py commits were initially update commits, but they were reduced to fixup commits during the rebase, as someone else upgraded them before me.

Depends on #187008

Close #235184

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->


_Fun fact:_ the shap module has been updated three times while marked broken.

pbsds committed:
python3Packages.ffmpy: init at 0.3.0

pbsds committed:
python3Packages.gradio: init at 3.20.1

evax reviewed (approved):
Thanks for the clear comments throughout!

pbsds commented:
@OfBorg build python3Packages.gradio

pbsds commented:
I moved the `python3Packages.shap` over to my other PR #187008, is it helps with testing.

SuperSandro2000 commented on pkgs/development/python-modules/analytics-python/default.nix:
```diff
@@ -0,0 +1,54 @@
+    sha256 = "plFBq25H2zlvW8Vwix25P/mpmILYH+gIIor9Xrtt/l8=";
+  };
+
+  # FIXME: use pythonRelaxDepsHook instead, likely blocked by https://github.com/NixOS/nixpkgs/issues/176695
+  # Relevant upstream PR: https://github.com/segmentio/analytics-python/pull/223
```
```suggestion
```
the hook is not for setup.py and not a requirement, only a convenience.

SuperSandro2000 commented on pkgs/development/python-modules/analytics-python/default.nix:
```diff
@@ -0,0 +1,54 @@
+
+  # FIXME: use pythonRelaxDepsHook instead, likely blocked by https://github.com/NixOS/nixpkgs/issues/176695
+  # Relevant upstream PR: https://github.com/segmentio/analytics-python/pull/223
+  prePatch = ''
```
```suggestion
  postPatch = ''
```

SuperSandro2000 commented on pkgs/development/python-modules/ffmpy/default.nix:
```diff
@@ -0,0 +1,33 @@
+  disabled = pythonOlder "3.6";
+
+  # The github repo has no release tags, the pypi distribution has no tests.
+  # This package is quite trivial anyway, and the tests mainly play around with the ffmpeg cli interface.
```
```suggestion
  # This package is quite trivial anyway, and the tests mainly play around with the ffmpeg cli interface.
  # https://github.com/Ch00k/ffmpy/issues/60
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+, aiohttp
+, fastapi
+, ffmpy
+, markdown-it-py#[linkify,plugins]
```
```suggestion
, markdown-it-py
```

Please use optional-dependencies

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+  propagatedBuildInputs = [
+    analytics-python
+    aiohttp
+    # h11 # not needed, https://github.com/gradio-app/gradio/pull/1680
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+    fastapi
+    ffmpy
+    markdown-it-py
+  ] ++ markdown-it-py.optional-dependencies.plugins
+    ++ markdown-it-py.optional-dependencies.linkify
+    ++ [
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+    fsspec
+    httpx
+    pydantic
+  ];
```
```suggestion
  ] ++ markdown-it-py.optional-dependencies.plugins
  ++ markdown-it-py.optional-dependencies.linkify;
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+    scikitimage
+    shap
+    ipython
+   ];
+  disabledTestPaths = [
+    # touches network
+    "test/test_tunneling.py"
+    "test/test_external.py"
+  ];
```
```suggestion
   ];

  disabledTestPaths = [
    # requires network connection
    "test/test_tunneling.py"
    "test/test_external.py"
  ];

```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,140 @@
+    "test/test_external.py"
+  ];
+  disabledTests = [
+    # touches network
```
```suggestion
    # requires network connection
```
because local networking would work in tests

SuperSandro2000 commented on pkgs/development/python-modules/markdown-it-py/default.nix:
```diff
@@ -27,19 +28,31 @@ buildPythonPackage rec {
 
-  nativeBuildInputs = [
-    flit-core
+  patches = [
+    # Allow linkify-it-py v2.
+    # This only affects projects that depend on 'markdown-it-py[linkify]'
+    (fetchpatch {
+      # https://github.com/executablebooks/markdown-it-py/pull/218
+      url = "https://github.com/executablebooks/markdown-it-py/commit/96931bbc2afa4aafb321e24e64af20ec14a7187f.patch";
+      name = "hukkin-allow-linkify-v2.patch";
+      sha256 = "SqOtM0PXd8ZtSurqCqNEbJd9XoEAKl8B4D7kSC3Fq/Q=";
+    })
   ];
```
Please use substituteInPlace to prevent apply errors in the future.

pbsds replied:
It will also error?

SuperSandro2000 replied:
It will not error and also not complain if the surrounding lines change.

pbsds commented:
@ofborg build  python3Packages.gradio

SuperSandro2000 commented on pkgs/development/python-modules/markdown-it-py/default.nix:
```diff
@@ -27,19 +27,31 @@ buildPythonPackage rec {
-  nativeBuildInputs = [
-    flit-core
-  ];
+  # Allow linkify-it-py v2.
+  # This only affects projects that depend on 'markdown-it-py[linkify]'
+  # https://github.com/executablebooks/markdown-it-py/pull/218
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace \
+        'linkify = ["linkify-it-py~=1.0"]' \
+        'linkify = ["linkify-it-py>=1,<3"]'
```
```suggestion
  postPatch = ''
    # Allow linkify-it-py v2.
    # This only affects projects that depend on 'markdown-it-py[linkify]'
    # https://github.com/executablebooks/markdown-it-py/pull/218
    substituteInPlace pyproject.toml \
      --replace "linkify-it-py~=1.0" "linkify-it-py>=1,<3"
```

pbsds commented:
This force-push changes the following:

* `python3Packages.hatch-requirements-txt`: init at 0.2.0
* `python3Packages.gradio`: 3.1.4 -> 3.11.0, migrate to hatchling, automate test selection

Gradio have now moved to pytest, which means i could steal the pytest hook i wrote for `shap` which automatically ignores failing network tests. I additionally expanded it to ignore tests that fail as a result of gradio upstream failing to upload certain test files to pypi. 

_Idea_: Should i generalize turn this pytest xfail hook into a a `checkInputs` hook like `pytestCheckHook`? I believe other maintainers would love to have something like this.

SuperSandro2000 commented on pkgs/development/python-modules/hatch-requirements-txt/default.nix:
```diff
@@ -0,0 +1,36 @@
+  # and its releases are also more frequent than github tags
+  src = fetchPypi {
+    inherit version;
+    pname = builtins.replaceStrings ["-"] ["_"] pname;
```
```suggestion
    pname = lib.replaceStrings ["-"] ["_"] pname;
```

pbsds commented on pkgs/development/python-modules/hatch-requirements-txt/default.nix:
```diff
@@ -0,0 +1,36 @@
+  ];
+
+  pythonImportsCheck = [
+    (builtins.replaceStrings ["-"] ["_"] pname)
```
For sake of consistency I'd rather pick one of the two suggested fixes than to do two different things. Which one do you prefer?

SuperSandro2000 replied:
Which suggested fixes you mean?

We never substitute pythonImportsCheck because it is pure coincidence that it matches pname at all and things likely break when normalizing pname.

SuperSandro2000 commented:
> _Idea_: Should i generalize turn this pytest xfail hook into a a `checkInputs` hook like `pytestCheckHook`? I believe other maintainers would love to have something like this.

Lets move that into another PR to get this one merged faster.

suhr commented:
I really hope it gets merged soon.

pbsds commented:
Rebased the pr, seems like the shap dependency broke again,

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,192 @@
+, markdown-it-py
+, mdit-py-plugins
+, markupsafe
+#, linkify-it-py
+#, mdit-py-plugins
```
please remove commented inputs

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,192 @@
+    numpy
+    orjson
+    pandas
+    #paramiko
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,192 @@
+   ];
+
+  # Add pytest hook skipping tests that access network, marking them as "Expected fail" (xfail)
+  # We additionally xfail FileNotFoundError, as gradio often fail to upload test files to pypi.
```
Why not fetch from github than instead?

pbsds replied:
The pypi build includes prebuildt web assets

SuperSandro2000 replied:
Could we fetch them separately? Python is very brittle and we really would like to run tests

SuperSandro2000 commented on pkgs/development/python-modules/gradio/default.nix:
```diff
@@ -0,0 +1,192 @@
+    # makes pytest freeze 50% of the time
+    "test/test_interfaces.py"
+  ];
+  #pytestFlagsArray = [ "-x" "-W" "ignore" ]; # uncomment me for debugging help
```
Does that spam log a lot? If not I would like to have it on by default to have the information straight in the logs.

pbsds replied:
It hides warnings and aborts on first fail

pbsds replied:
This also stops on first error. The tests are a bit slow, and the output is basically unreadable with many failiures.

Janik-Haag reviewed (commented):
Can you apply the patches Sandro suggested so we can get this merged?

pbsds committed:
python3Packages.markdown-it-py: fix linkify extra, switch to flit format

linkify-it-py is on version 2.0.0, while markdown-it-py[linkify] which depends on linkify-it-py~=1.0

pbsds commented:
I'm tired of waiting for shap to fail building. Not fixing it anymore, i'll just disable the relevant test if it is broken.

Not bumping gradio version this time either, since they have done major changes which introduces a chicken-and-the-egg circular dependency problem. I'll solve that in a follow-up pr along with possibly mixing the github and pypi sources.

@ofborg build python3Packages.gradio

Janik-Haag commented:
I would merge this but x86_64-darwin  fails, can you fix it or mark it as broken?

pbsds commented:
Seems like tensorflow timed out? Not sure what to do about that other than trying again

@ofborg build python3Packages.gradio

pbsds commented:
This time darwin succeeded while x86_64-linux instead failed, on the same package: scikit-image. Likely that package is flaky.
It's pretty safe to assume the problem is not introduced in this PR.

How do i rebase to staging?

Janik-Haag commented:
> How do i rebase to staging?

Isn't changing the target branch of the pr to staging enough?
https://nixos.org/manual/nixpkgs/stable/#id-1.7.4.8.3

happysalada commented:
something like
```
git fetch upstream staging:staging
git checkout staging
git checkout -b gradio_staging
git cherry-pick gradio-init~2 gradio-init~1 gradio-init~0
```
looking forward to this, feel free to post if you need anything.

happysalada commented:
The build gets stuck on linux. But here are some of the failures
```
  ├─ ⚠ pacup failed with exit code 1 after ⏱ 7s in installPhase
┃    ├─ ⚠ python3.10-returns-0.20.0 failed with exit code 1 after ⏱ 20m43s in pytestCheckPhase
┃    ├─ ⚠ matrix-synapse-1.86.0 failed with exit code 1 after ⏱ 12m25s in installCheckPhase
┃    ├─ ⚠ python3.10-qcodes-0.38.1 failed with exit code 1 after ⏱ 17m44s in pytestCheckPhase
┃    ├─ ⚠ python3.10-spyder-5.4.2 failed with exit code 1 after ⏱ 36s in installPhase
┃    ├─ ⚠ python3.10-numpyro-0.11.0 failed with exit code 139 after ⏱ 21m48s in pytestCheckPhase
┃    ├─ ⚠ python3.10-nbconflux-0.7.0 failed with exit code 1 after ⏱ 2s in buildPhase
┃    ├─ ⚠ python3.10-ocrmypdf-14.2.1 failed with exit code 1 after ⏱ 6m12s in pytestCheckPhase
┃    ├─ ⚠ python3.10-datasette-0.64.3 failed with exit code 1 after ⏱ 1m33s in pytestCheckPhase
┃    ├─ ⚠ conan-2.0.5 failed with exit code 1 after ⏱ 12m18s in pytestCheckPhase
┃    ├─ ⚠ python3.11-camelot-py-0.11.0 failed with exit code 1 after ⏱ 1s in buildPhase
```

the build is stuck for me on the following
```
 manim-0.16.0.post0 (pytestCheckPhase) ⏱ 3h27m40s
┃    ├─ ⏵ python3.10-tables-3.8.0 (installCheckPhase) ⏱ 3h48m14s
┃    ├─ ⏵ python3.10-stim-1.9.0 (buildPhase) ⏱ 4h2m42s
┃    ├─ ⏵ python3.10-pytorch-metric-learning-2.2.0 (pytestCheckPhase) ⏱ 4h5m52s
┃    ├─ ⏵ python3.10-botorch-0.8.5 (pytestCheckPhase) ⏱ 4h14m21s
┃    ├─ ⏵ miopen-5.4.2 (buildPhase) ⏱ 4h53m32s
```

pbsds commented:
I hoped there would be some bot to do it or something

@ofborg build python3Packages.gradio

flokli commented:
It looks like you accidentally mass-pinged a bunch of people, which are now subscribed
and getting notifications for everything in this pull request. Unfortunately, they
cannot be automatically unsubscribed from the issue (removing review request does not
unsubscribe), therefore development cannot continue in this pull request anymore.

Please open a new pull request with your changes, link back to this one and ping the
people actually involved in here over there.

In order to avoid this in the future, there are instructions for how to properly
rebase between branches in our [contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging).
Setting your pull request to draft prior to rebasing is strongly recommended.
In draft status, you can preview the list of people that are about to be requested
for review, which allows you to sidestep this issue.
This is not a bulletproof method though, as OfBorg still does review requests even on draft PRs.
